New terrain analysis and modeling tutorial (#71)
OSGeo/grass-tutorials: content/tutorials/terrain_and_DEMs/GRASS_terrain.qmd

---
title: "Visualizing and Modeling Terrain from DEMs in GRASS"
author: "Eunice Villaseñor Iribe and Michael Barton"
date: 2025-06-06
date-modified: today
lightbox: true
image: img_terrain/thumbnail.webp
format: 
  html:
    embed-resources: true
    toc: true
    code-tools: true
    code-copy: true
    code-fold: false
page-layout: article

> categories: [beginner, intermediate, raster, DEM, visualization, terrain, topography, slope, aspect, landforms, stream networks, watershed]

Reviewer: And what about the topography terms?

There is a discussion in #73.

> categories: [beginner, intermediate, raster, DEM, visualization, terrain, topography, slope, aspect, landforms, stream networks, watershed]

Author: I just saw this. Do you think topography is redundant with terrain? Which would most people want to look for? Terrain is in the title, so it is not needed for search, only for grouping. I could go either way. What do the others think?

description: This tutorial guides a user through multiple ways of analyzing and modeling terrain from DEM raster maps.
execute: 
  eval: false
copyright: 
  holder: Eunice Villaseñor Iribe and Michael Barton
  year: 2025
funding: "Creation of this tutorial was supported in part by US National Science Foundation grant FAIN 2303651."
---

GRASS has many tools that can help you visualize and model terrain from digital elevation model (DEM) raster maps. A few examples include:  

-  [r.shade](https://grass.osgeo.org/grass-stable/manuals/r.shade.html) for visualization,  

-  [r.slope.aspect](https://grass.osgeo.org/grass-stable/manuals/r.slope.aspect.html) for modeling slope and aspect,  

-  [r.geomorphon](https://grass.osgeo.org/grass-stable/manuals/r.geomorphon.html) for modeling topographic features,  

-  [r.watershed](https://grass.osgeo.org/grass-stable/manuals/r.watershed.html) for modeling overland flow and watershed basins, and

> -  [r.stream.extract](https://grass.osgeo.org/grass%20stable/manuals/r.stream.extract.html) to generate stream networks.  

Reviewer: ```suggestion
-  [r.stream.extract](https://grass.osgeo.org/grass-stable/manuals/r.stream.extract.html) to generate stream networks.  
```

This tutorial will explore these tools, but they are only a few of the many GRASS modules to represent and analyze landscapes.  

::: {.callout-note title="Demo Dataset"}
This tutorial uses one of the standard GRASS sample data sets for Flagstaff, Arizona, USA: ***flagstaff_arizona_usa***. We will refer to place names in that data set, but it can be completed with any of the [standard sample data sets](https://grass.osgeo.org/download/data/) for any region--for example, the [North Carolina data set](https://grass.osgeo.org/sampledata/north_carolina/nc_spm_08_grass7.zip). We will use the *elevation* DEM.  
:::

> ::: {.callout-note title="GRASS environment"}

Reviewer: ```suggestion
::: {.callout-note title="GRASS interfaces"}
```

## Applying a color table to a DEM  

GRASS offers multiple ways to apply color to a DEM to visualize terrain under the Raster/Manage colors menu. The simplest way is to apply a color table using the [r.colors](https://grass.osgeo.org/grass-stable/manuals/r.colors.html) module.  

Let's apply the *elevation* color table to the Flagstaff DEM. This color table assigns 'hot' colors to higher elevations and 'cool' colors to lower elevations.  

::::::::: {.panel-tabset group="language"}

#### GUI

::::: grid
::: g-col-6

1.  Select [r.colors](https://grass.osgeo.org/grass-stable/manuals/r.colors.html) tool from the Raster/Manage colors menu.  

2.  In the "Map" tab, select *elevation* map in the "Name of raster map(s)" entry box.
:::

::: g-col-6
![](img_terrain/r_colors1.webp)  
:::
:::::

::::: grid
::: g-col-6
3.  In the Define tab, select *elevation* as the color table.  

4.  Click Run.  
:::

::: g-col-6
![](img_terrain/r_colors2.webp)
:::
:::::

Try out some other color tables. ETOPO2, terrain, and the SRTM color tables are also good for terrain colors. You can also make a custom color table easily, described in the r.colors manual.  

#### Command line

> ```{python}

Author: OK. I'll try this.

This approach can also be used to visualize other terrain analyses completed later in this tutorial.  

:::::::::::: {.panel-tabset group="language"}

#### GUI

We will start by creating a new color-shaded relief map using [r.shade](https://grass.osgeo.org/grass-stable/manuals/r.shade.html).  

::::: grid
::: g-col-6

1.  Open *r.shade* from the Raster/Terrain analysis/Apply shade to raster menu.  

2.  Set "Name of shaded relief" map to to *relief*.  

3.  Set "Name of raster to drape over relief raster" to *elevation*.  

4.  Enter *topography_colored_relief* as "Name of shaded raster" map.  
:::

::: g-col-6
![](img_terrain/r_shade1.webp)
:::
:::::

::::: grid
::: g-col-6
5.  Under the "Optional" tab, set Brightness to 40% to provide a more colorful result.

6.  Click Run.  
:::

::: g-col-6
![](img_terrain/r_shade2.webp)  
:::
:::::

#### Command line  

```{python}
r.shade shade=relief color=elevation output=topography_colored_relief brighten=40]
```

#### Python

```{python}
gs.run_command("r.shade", 
                shade="relief, 
                color="elevation", 
                output="topography_colored_relief", 

>                 brighten=40)

Author: I've formatted the Python code as above throughout. If it is truly a problem I'll change them all. Otherwise, I'll leave this one as is to be consistent with the other code.

-  While there are many possible inputs to affect how *r.watershed* models different hydrological characteristics, we will use the simplest example here.  

-  The minimum required for modeling watershed basins, flow accumulations, and stream networks is a DEM (*elevation*) and the minimum size of the watershed basins to model (measured in number of cells). We will create basins with a minimum size of **10,000** cells (9 km2). Choosing a smaller number would generate a map of many smaller basins.  

-  Although we start with mapping watershed basins, to save time, we will have *r.watershed* also generate outputs for flow accumulation and stream segments at the same time is is creating a basin map.  

-  To produce better looking basin and streams maps, we will also select the "Positive flow accumulation" and "Beautify flat areas" options.  

::: {.callout-note title="Demo Dataset"}
*r.watershed* runs very fast for this kind of modeling. Nevertheless, it is advisable to give it as much RAM as you can spare under the Optional tab.  
:::

::::::::::::::: {.panel-tabset group="language"}

#### GUI

::::: grid
::: g-col-6

> 1.  Open the [r.watershed](https://grass.osgeo.org/grass84/manuals/r.watershed.html) module from the Raster/Hydrologic modeling/Watershed analysis menu.  

Reviewer: ```suggestion
1.  Open the [r.watershed](https://grass.osgeo.org/grass-stable/manuals/r.watershed.html) module from the Raster/Hydrologic modeling/Watershed analysis menu.  
```

                basin="watershed_basin10000", 
                stream="watershed_streams10000", 
                flags="ab")

```

:::::::::::::::

Resulting map of watershed basins.  

![watershed basins with ≥ 10,000 cells coloring a relief map](img_terrain/watershed_basins.webp)  

## Flow accumulation  

Above, we also made a flow accumulation map. A flow accumulation map models the potential flow of water over a terrain:  

::::: grid
::: g-col-6

-  Assuming a centimeter of rain falls on each cell of a DEM, the values of a flow accumulation map will show the depth of water in centimeters that accumulates and flows over each cell as the water flows into the cell from adjacent upslope cells.  

-  The map we made assumes that the same amount of rain fell on every cell and topographic relief is the only thing affecting overland flow of water. Hence, the flow accumulation model generated represents maximal values.  

-  Additional settings in *r.watershed* enable modeling overland flow more realistically with different amounts of rain falling across the region, and vegetation, soil characteristics, or barriers like dams or dikes also affecting accumulation.  

> -  [r.watershed](https://grass.osgeo.org/grass84/manuals/r.watershed.html) uses a **multiple flow direction** algorithm by default in which water flowing out of a cell is distributed among all adjacent downslope cells. The greatest flow is directed to the adjacent downslope cell with the greatest elevation difference and lower flows are allocated to downslope cells with lesser elevation difference.  

Reviewer: It may be relevant to point here to the fact that the algortihm doesn't need to have sinks filled first, but it uses LCP to route water out of sinks.

Author: Just saw this (was hidden). Will add.

> -  [r.watershed](https://grass.osgeo.org/grass84/manuals/r.watershed.html) uses a **multiple flow direction** algorithm by default in which water flowing out of a cell is distributed among all adjacent downslope cells. The greatest flow is directed to the adjacent downslope cell with the greatest elevation difference and lower flows are allocated to downslope cells with lesser elevation difference.  

Reviewer: ```suggestion
-  [r.watershed](https://grass.osgeo.org/grass-stable/manuals/r.watershed.html) uses a **multiple flow direction** algorithm by default in which water flowing out of a cell is distributed among all adjacent downslope cells. The greatest flow is directed to the adjacent downslope cell with the greatest elevation difference and lower flows are allocated to downslope cells with lesser elevation difference.  
```

:::
:::::

As with slope, a legend can help convey the information of a flow accumulation map more clearly. The distribution of accumulation values is non-linear, however, with many cells on hillslopes having minimal accumulation and a few cells in the bottoms of major streams having much higher values. For this reason, a logarithmic scale better communicates the relationships between colors and values. Create a legend as described previously but with some additional options.

::::::::::::::: {.panel-tabset group="language"}

#### GUI

-  Name of raster map: watershed_accumulation10000@terrain (Input tab).  

-  Legend title: "Flow Accumulation" (Title tab).  

-  Title font size: 14 (Title tab).  

-  Use logarithmic scale: check the box (Advanced tab).  

-  Specific values to draw ticks: 1,10,100,1000,10000,100000,1000000 (Gradient tab).  

-  Font size: 12 (Font settings tab).  

#### Command line

```{python}
d.legend -l raster=watershed_accumulation100000 title="Flow Accumulation" title_fontsize=14 label_values=1,10,100,1000,10000,100000,1000000 fontsize=12
```

#### Python

```{python}

> gs.run_command("d.legend", 
>                 raster="watershed_accumulation10000", 
>                 title="Flow Accumulation", 
>                 title_fontsize=14, 
>                 label_values="1,10,100,1000,10000,100000,1000000", 
>                 fontsize=12, 
>                 flags="l")

Reviewer: ```suggestion
gs.run_command(
    "d.legend",
    raster="watershed_accumulation10000",
    title="Flow Accumulation",
    title_fontsize=14,
    label_values="1,10,100,1000,10000,100000,1000000",
    fontsize=12,
    flags="l",
)
```

```

:::::::::::::::

Here is a flow accumulation map with a legend using a logarithmic scale. Note higher accumulation on the slopes of the San Francisco peaks, in major stream courses, and in small, shallow lake depressions.  

![flow accumulation map coloring a relief map with logarithmic scale legend](img_terrain/watershed_accumulation_legend.webp)

## Stream networks

*r.watershed* also produced a maps of streams, representing the primary drainages of each 10,000-cell watershed basin.  

::::: grid
::: g-col-6

-  It is a raster map of streams with each stream cell coded and colored to match the basin that it drains, seen in the map of the Flagstaff region to the right.  

-  The background is set to black in the display properties because it is otherwise very difficult to distinguish the raster stream cells.  
:::

::: g-col-6
![raster stream segments map for each watershed basin](img_terrain/watershed_streams.webp)  
:::
:::::

### Vectorizing stream network with *r.to.vect*  

> We can make the streams more easily visible if we convert the raster stream map to a vector line map. This can be done easily with the [r.to.vect](https://grass.osgeo.org/grass84/manuals/r.to.vect.html) tool that converts rasters maps into vectors.  

Reviewer: ```suggestion
We can make the streams more easily visible if we convert the raster stream map to a vector line map. This can be done easily with the [r.to.vect](https://grass.osgeo.org/grass-stable/manuals/r.to.vect.html) tool that converts rasters maps into vectors.  
```

-  *r.to.vect* can convert points, lines, and areas. In this case, we will convert the stream segments created by *r.watershed* into vector lines.  

-  The raster stream segments are coded with the values of the watershed basins they drain. These values can be transferred to the equivalent vector line segments.  

::::::::::::::: {.panel-tabset group="language"}

#### GUI

::::: grid
::: g-col-6

> 1.  Select the [r.to.vect](https://grass.osgeo.org/grass84/manuals/r.to.vect.html) tool from either the Raster/Map type conversions/Raster to vector **or** the File/Map type conversions/Raster to vector menu.  

Reviewer: ```suggestion
1.  Select the [r.to.vect](https://grass.osgeo.org/grass-stable/manuals/r.to.vect.html) tool from either the Raster/Map type conversions/Raster to vector **or** the File/Map type conversions/Raster to vector menu.  
```

2.  Enter the map *watershed_streams10000* as the "Name of input raster map".  

3.  Enter *watershed_streams10000* as the "Name for output vector map".  

4.  Enter *line* as the "Output feature type".  

5.  Under the optional tab, check the box for "Use raster values as categories instead of unique sequence". This will copy the basin number into the *CAT* index key field of the new vector line map.  

6.  Click Run.  
:::

::: g-col-6
![](img_terrain/r.vect1.webp)  
:::
:::::

#### Command line

```{python}
r.to.vect -v input=watershed_streams10000 output=watershed_streams10000 type=line
```

#### Python

```{python}

> gs.run_command("r.to.vect", 
>                 input="watershed_streams10000", 
>                 output="watershed_streams10000", 
>                 type="line",
>                 flags="v")

Reviewer: ```suggestion
gs.run_command(
    "r.to.vect",
    input="watershed_streams10000",
    output="watershed_streams10000",
    type="line",
    flags="v",
)
```

```

:::::::::::::::

The maps below show the vector stream network overlaying watershed basins on the left and overlaying a color shaded relief map on the right.  

::::: grid
::: g-col-6
![vector streams and watershed basins](img_terrain/watershed_basins&streamvectors.webp)  
:::

::: g-col-6
![vector streams and topographic relief](img_terrain/relief&streamvectors.webp)  
:::
:::::

::: {.callout-note title="Tip"}

> For a more detailed stream network, run *r.watershed* with a minimum basin size smaller than the 10,000 cells used above. If you choose a very small minimum basin size to create a very detailed stream map, you may be prompted to run [r.thin](https://grass.osgeo.org/grass84/manuals/r.thin.html) before running *r.to.vect* to make sure that raster stream map consists of only single grid cell stream lines. 

Reviewer: ```suggestion
For a more detailed stream network, run *r.watershed* with a minimum basin size smaller than the 10,000 cells used above. If you choose a very small minimum basin size to create a very detailed stream map, you may be prompted to run [r.thin](https://grass.osgeo.org/grass-stable/manuals/r.thin.html) before running *r.to.vect* to make sure that raster stream map consists of only single grid cell stream lines. 
```

:::

### Vectorizing stream network with *r.stream.extract*  

> An alternative way to generate a stream network from a watershed analysis is with the [r.stream.extract](https://grass.osgeo.org/grass84/manuals/r.stream.extract.html) tool.  

Reviewer: ```suggestion
A better way to generate a stream network from a watershed analysis is with the [r.stream.extract](https://grass.osgeo.org/grass-stable/manuals/r.stream.extract.html) tool.  
```

2.  Enter *elevation* for the "Name of input elevation raster map".  

3.  Enter 5000 for "Minimum flow accumulation for streams" (smaller numbers will generate a more detailed stream network, while larger numbers will generate a less detailed network).  
:::

::: g-col-6
![](img_terrain/r.stream.extract1.webp)  
:::
:::::

::::: grid
::: g-col-6
4.  Enter *watershed_accumulation10000* for the "Name of input accumulation raster map".  
:::

::: g-col-6
![](img_terrain/r.stream.extract2.webp)  
:::
:::::

::::: grid
::: g-col-6
5.  Enter *streams5000accumulation* for "Name for output vector map with unique stream id".  

6.  Click Run.  
:::

::: g-col-6
![](img_terrain/r.stream.extract3.webp)  
:::
:::::

#### Command line

```{python}
r.stream.extract elevation=elevation accumulation=watershed_accumulation10000@ threshold=5000 stream_vector=streams5000accumulation
```

#### Python

```{python}
gs.run_command("r.stream.extract", 
                elevation="elevation", 
                accumulation="watershed_accumulation10000@", 
                threshold=5000, 
                stream_vector="streams5000accumulation")
```

:::::::::::::::

> The 2 images below show the stream network generated by *r.stream.extract* overlaying a color shaded relief map (left) and a comparison of this network with the one generated by r.watershed and then vectorized with *r.to.vect(* (right).  

Reviewer: ```suggestion
The 2 images below show the stream network generated by *r.stream.extract* overlaying a color shaded relief map (left) and a comparison of this network with the one generated by r.watershed and then vectorized with *r.to.vect* (right).  
```

> -  Note the stream lines (blue) and points (red) indicating stream heads and junctions in the left image. Lines or points can be displayed or hidden or shown together in the vector properties tool [d.vect](https://grass.osgeo.org/grass84/manuals/d.vect.html).  

Reviewer: ```suggestion
-  Note the stream lines (blue) and points (red) indicating stream heads and junctions in the left image. Lines or points can be displayed or hidden or shown together in the vector properties tool [d.vect](https://grass.osgeo.org/grass-stable/manuals/d.vect.html).  
```

-  In the image to the right, the more detailed stream network generated by *r.stream.extract* is shown in narrow blue lines and the the stream network generated by *r.watershed* and *r.to.vect* is shown as wider yellow lines.  

::::: grid
::: g-col-6
![](img_terrain/relief&extracted_streams.webp)  
:::

::: g-col-6
![](img_terrain/stream_comparison.webp)  
:::
:::::

::: {.callout-note title="Tip"}
To display the stream network without the junction nodes, uncheck the "point" box in the Selection tab of the vector display properties tool (command: *d.vect map=streams5000accumulation type=line*)
:::

# Summary

> In this tutorial we have gone over only a few the many tools available in GRASS to conduct terrain and watershed analysis. We used the Flagstaff sample dataset to complete this analysis, but you can also try this other datasets and DEM maps. You can also use the GRASS sample datasets to explore the Many other tools available under the terrain analysis and [hydrology](https://grass.osgeo.org/grass84/manuals/topic_hydrology.html) hydrology sections. In addition to the main GRASS modules, there are also many others available as [GRASS addons](https://grass.osgeo.org/grass-stable/manuals/addons/).  

Reviewer: ```suggestion
In this tutorial we have gone over only a few of the many tools available in GRASS to conduct terrain and watershed analysis. We used the Flagstaff sample dataset to complete this analysis, but you can also try this other datasets and DEM maps. You can also use the GRASS sample datasets to explore the many other tools available under the terrain analysis and [hydrology](https://grass.osgeo.org/grass-stable/manuals/topic_hydrology.html) hydrology sections. In addition to the main GRASS modules, there are also many others available as [GRASS addons](https://grass.osgeo.org/grass-stable/manuals/addons/).  
```

2.  Set "Name of elevation raster map" to *elevation*.  

3.  Enter 10,000 for "Minimum size of exterior watershed basin".  
:::

::: g-col-6
![](img_terrain/r.watershed1.webp)  
:::
:::::

::::: grid
::: g-col-6
4.  Under the Outputs tab, enter:  

-  watershed_accumulation10000 for "accumulation",  

-  watershed_basin10000 for "basins", and  

-  watershed_streams10000 for "stream segments".  
:::

::: g-col-6
![](img_terrain/r.watershed2.webp)
:::
:::::

::::: grid
::: g-col-6
5.  Under the Optional tab:  

-  Check the "Positive flow accumulation" box. Otherwise flow accumulation will be negative for basins at the map edge, to indicate that they may under-represent accumulation.  

-  Check the "Beautify flat areas" box to have more accurate stream segments across flat areas.  

-  Modify available memory as appropriate for your system.  

6.  Click Run  
:::

::: g-col-6
![](img_terrain/r.watershed3.webp)  
:::
:::::

#### Command line

```{python}
r.watershed -a -b elevation=elevation threshold=10000 accumulation=watershed_accumulation10000 basin=watershed_basins10000 stream=watershed_streams10000
```

#### Python

```{python}
gs.run_command("r.watershed", 

>                 elevation="elevation\@TerrainAnalysis", 

Author: Thanks. I'll go through and make the corrections.
